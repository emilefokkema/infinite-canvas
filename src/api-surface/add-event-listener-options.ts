/**
 * Options that can be provided when adding an event listener to an {@link InfiniteCanvas}
 */
export interface AddEventListenerOptions{
    /**
     * If `true`, the listener will be invoked at most once
     */
    once: boolean;
}