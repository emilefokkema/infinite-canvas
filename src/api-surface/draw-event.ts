import { Transformed } from "./transformed";

export interface DrawEvent extends Transformed{
}
