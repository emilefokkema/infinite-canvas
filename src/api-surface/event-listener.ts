import { EventMap } from "./event-map";


export declare type EventListener<K extends keyof EventMap> = (ev: EventMap[K]) => void;
