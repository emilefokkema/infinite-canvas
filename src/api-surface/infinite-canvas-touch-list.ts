import { InfiniteCanvasTouch } from "./infinite-canvas-touch";

export interface InfiniteCanvasTouchList extends TouchList{
    item(index: number): InfiniteCanvasTouch | null;
    [index: number]: InfiniteCanvasTouch;
}

export * from './infinite-canvas-touch';