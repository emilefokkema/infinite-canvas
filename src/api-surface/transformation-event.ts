import { Transformed } from "./transformed";

export interface TransformationEvent extends Event, Transformed{
}
