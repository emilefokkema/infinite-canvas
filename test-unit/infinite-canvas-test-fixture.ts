import { Mock, MockInstance, vi } from 'vitest'
import { DrawingLock } from "src/drawing-lock";
import { MockCanvasMeasurementProvider } from './mock-canvas-measurement-provider';
import { CanvasContextMock } from './canvas-context-mock';
import { Config } from 'api/config';
import { MockDrawingIterationProvider } from './mock-drawing-iteration-provider';
import { CssLengthConverterFactory } from 'src/css-length-converter-factory';
import { InfiniteCanvasViewBox } from 'src/infinite-canvas-viewbox';
import { RectangleManagerImpl } from 'src/rectangle/rectangle-manager-impl';
import { InfiniteContext } from 'src/infinite-context/infinite-context';
import { ViewBox } from 'src/interfaces/viewbox';
import { InfiniteCanvasRenderingContext2D } from 'api/infinite-canvas-rendering-context-2d';


function setupGlobals(){
	window.createImageBitmap = function(): Promise<ImageBitmap>{return undefined;};
	(<any>window).ImageData = class {
		public width: number;
		public height: number;
		public data: Uint8ClampedArray;
		constructor(arrayOrWidth: Uint8ClampedArray | number, widthOrHeight: number, height?: number){
			if(typeof arrayOrWidth === "number"){
				this.width = arrayOrWidth;
				this.height = widthOrHeight;
			}else{
				this.width = widthOrHeight;
				this.height = height;
				this.data = arrayOrWidth;
			}
		}
	};
}

export interface InfiniteCanvasTestFixture{
    width: number
    height: number
    releaseDrawingLockSpy: MockInstance<()=>void>
    getDrawingLockSpy: MockInstance<()=>DrawingLock>
    measurementProvider: MockCanvasMeasurementProvider
    contextMock: CanvasContextMock
    config: Partial<Config>
    mockDrawingIterationProvider: MockDrawingIterationProvider
    viewbox: ViewBox,
    infiniteContext: InfiniteCanvasRenderingContext2D
}

export function createInfiniteCanvasTestFixture(): InfiniteCanvasTestFixture{
    setupGlobals();
    const drawingLock: DrawingLock = {release(){}};
    const releaseDrawingLockSpy = vi.spyOn(drawingLock, 'release');
    const getDrawingLockSpy = vi.fn<()=>DrawingLock>().mockReturnValue(drawingLock)
    const width = 200;
    const height = 200;
    const measurementProvider = new MockCanvasMeasurementProvider(width, height);
    const contextMock = new CanvasContextMock();
    const config = {}
    const mockDrawingIterationProvider = new MockDrawingIterationProvider()
    const cssLengthConverterFactory: CssLengthConverterFactory = {
        create: () => ({
            getNumberOfPixels(value: number){
                return value;
            }
        })
    };
    const viewbox = new InfiniteCanvasViewBox(
        new RectangleManagerImpl(measurementProvider, config),
        contextMock.mock,
        mockDrawingIterationProvider,
        getDrawingLockSpy,
        () => false);
    const infiniteContext = new InfiniteContext(undefined, viewbox, cssLengthConverterFactory);
    return {
        width,
        height,
        releaseDrawingLockSpy,
        getDrawingLockSpy,
        measurementProvider,
        contextMock,
        config,
        mockDrawingIterationProvider,
        viewbox,
        infiniteContext
    };
}