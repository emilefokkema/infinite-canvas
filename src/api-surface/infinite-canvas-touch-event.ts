import { InfiniteCanvasTouchList } from "./infinite-canvas-touch-list";

export interface InfiniteCanvasTouchEvent extends TouchEvent{
    readonly targetTouches: InfiniteCanvasTouchList;
    readonly changedTouches: InfiniteCanvasTouchList;
    readonly touches: InfiniteCanvasTouchList;
}

export * from './infinite-canvas-touch-list';