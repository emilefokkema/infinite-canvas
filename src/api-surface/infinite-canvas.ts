import {Config} from "./config";
import { Units } from "./units";
import {InfiniteCanvasRenderingContext2D} from "./infinite-canvas-rendering-context-2d";
import {EventMap} from "./event-map";
import { TransformationEvent } from "./transformation-event";
import { Transformed } from './transformed';

export interface InfiniteCanvasEventHandlers extends DocumentAndElementEventHandlers, GlobalEventHandlers {
    /**
     * The [event handler property](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events#event_handler_properties) for the {@link EventMap.transformationstart} event
     */
    ontransformationstart: ((this: InfiniteCanvas, event: TransformationEvent) => any) | null;
    /**
     * The [event handler property](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events#event_handler_properties) for the {@link EventMap.transformationchange} event
     */
    ontransformationchange: ((this: InfiniteCanvas, event: TransformationEvent) => any) | null;
    /**
     * The [event handler property](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events#event_handler_properties) for the {@link EventMap.transformationend} event
     */
    ontransformationend: ((this: InfiniteCanvas, event: TransformationEvent) => any) | null;
    /**
     * The [event handler property](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events#event_handler_properties) for the {@link EventMap.draw} event
     */
    ondraw: ((this: InfiniteCanvas, event: TransformationEvent) => any) | null;
    /**
     * The [event handler property](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events#event_handler_properties) for the {@link EventMap.wheelignored} event
     */
    onwheelignored:  ((this: InfiniteCanvas, event: Event) => any) | null;
    /**
     * The [event handler property](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events#event_handler_properties) for the {@link EventMap.touchignored} event
     */
    ontouchignored:  ((this: InfiniteCanvas, event: Event) => any) | null;
    /**
     * See [`addEventListener`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener) 
     */
    addEventListener<K extends keyof EventMap>(type: K, listener: (this: InfiniteCanvas, ev: EventMap[K]) => any, options?: boolean | AddEventListenerOptions): void;
    addEventListener(type: string, listener: EventListenerOrEventListenerObject, options?: boolean | AddEventListenerOptions): void;
    /**
     * See [`removeEventListener`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/removeEventListener)
     */
    removeEventListener<K extends keyof EventMap>(type: K, listener: (this: InfiniteCanvas, ev: EventMap[K]) => any, options?: boolean | EventListenerOptions): void;
    removeEventListener(type: string, listener: EventListenerOrEventListenerObject, options?: boolean | EventListenerOptions): void;
}

export interface InfiniteCanvas extends Config, InfiniteCanvasEventHandlers, Transformed{
    /**
     * This methods return the {@link InfiniteCanvasRenderingContext2D} belonging to this instance of {@link InfiniteCanvas}
     * 
     * @param contextType for (partial) compatibility with the other [getContext()](https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/getContext)
     */
    getContext(contextType?: '2d'): InfiniteCanvasRenderingContext2D;
}

export interface InfiniteCanvasCtr {
    /**
     * Creates a new instance of an {@link InfiniteCanvas}
     * 
     * @param canvas The canvas on which to draw the {@link InfiniteCanvas}
     * @param config The {@link InfiniteCanvas}'s initial configuration
     */
    new(canvas: HTMLCanvasElement, config?: Partial<Config>): InfiniteCanvas;
    prototype: InfiniteCanvas;
    /**
     * Static property with a value of {@link Units.CANVAS}
     */
    CANVAS_UNITS: Units;
    /**
     * Static property with a value of {@link Units.CSS}
     */
    CSS_UNITS: Units;
};

declare var InfiniteCanvasConstructor: InfiniteCanvasCtr;

export * from './config'
export * from './event-map'
export * from './infinite-canvas-rendering-context-2d'
export * from './units'
export default InfiniteCanvasConstructor;
