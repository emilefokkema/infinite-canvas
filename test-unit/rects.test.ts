import { beforeEach, describe, expect, it} from 'vitest'
import {CanvasContextMock} from "./canvas-context-mock";
import {InfiniteCanvasRenderingContext2D} from "../src/api-surface/infinite-canvas-rendering-context-2d";
import { createInfiniteCanvasTestFixture } from './infinite-canvas-test-fixture';

describe('an infinite canvas context', () => {
	let infiniteContext: InfiniteCanvasRenderingContext2D;
	let contextMock: CanvasContextMock;

    beforeEach(() => {
		({contextMock, infiniteContext} = createInfiniteCanvasTestFixture());
	});

    describe("that fills a rect with finite width and positive infinite height", () => {

		beforeEach(() => {
			contextMock.clear();
			infiniteContext.fillRect(20, 20, 30, Infinity);
		});

		it("should fill a rect that extends to outside of the viewbox", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	});

    describe("that fills a rect with finite width and negative infinite height", () => {

		beforeEach(() => {
			contextMock.clear();
			infiniteContext.fillRect(20, 20, 30, -Infinity);
		});

		it("should fill a rect that extends to outside of the viewbox", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	});

    describe("that begins a path", () => {

		beforeEach(() => {
			infiniteContext.beginPath();
		});

		it("should get an error if it tries to add a rect with x and y that do not determine a direction", () => {
			expect(() => {
				infiniteContext.rect(-Infinity, -Infinity, Infinity, Infinity);
			}).toThrow();
		});

		it("should get an error if it tries to add a round rect with x and y that do not determine a direction", () => {
			expect(() => {
				infiniteContext.roundRect(-Infinity, -Infinity, Infinity, Infinity, 2);
			}).toThrow();
		});

		describe("and then adds a rect that has no area", () => {

			beforeEach(() => {
				infiniteContext.rect(-Infinity, 100, 100, 100);
			});

			describe("and then adds a line to a finite point and strokes it", () => {

				beforeEach(() => {
					infiniteContext.lineTo(100, 100);
					contextMock.clear();
					infiniteContext.stroke();
				});

				it("should have drawn a ray from the position of the rect without area", () => {
					expect(contextMock.getLog()).toMatchSnapshot();
				});
			});
		});
	});

    describe("that fills a rect with only a top edge", () => {

		beforeEach(() => {
			contextMock.clear();
			infiniteContext.fillRect(-Infinity, 100, Infinity, Infinity);
		});

		it("should fill a rect that fills the bottom half of the viewbox", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	});

    describe("that fills a rect with only a right edge", () => {

		beforeEach(() => {
			contextMock.clear();
			infiniteContext.fillRect(100, -Infinity, -Infinity, Infinity);
		});

		it("should fill a rect that fills the left half of the viewbox", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	});
    
    describe("that fills a rect without top edge with positive infinite height and a finite width", () => {

		beforeEach(() => {
			contextMock.clear();
			infiniteContext.fillRect(20, -Infinity, 30, Infinity);
		});

		it("should fill a rect that extends to the top and bottom of the viewbox", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	});

    describe("that fills a rect without bottom edge with negative infinite height and a finite width", () => {

		beforeEach(() => {
			contextMock.clear();
			infiniteContext.fillRect(20, Infinity, 30, -Infinity);
		});

		it("should fill a rect that extends to the top and bottom of the viewbox", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	});

    describe("that fills a rect without left edge with positive infinite width and a finite height", () => {

		beforeEach(() => {
			contextMock.clear();
			infiniteContext.fillRect(-Infinity, 20, Infinity, 30);
		});

		it("should fill a rect that extends to the left and right of the viewbox", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	});

    describe("that fills a rect without right edge with negative infinite width and a finite height", () => {

		beforeEach(() => {
			contextMock.clear();
			infiniteContext.fillRect(Infinity, 20, -Infinity, 30);
		});

		it("should fill a rect that extends to the left and right of the viewbox", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	});

    describe("that fills a rect with positive infinite width and a finite height", () => {

		beforeEach(() => {
			contextMock.clear();
			infiniteContext.fillRect(20, 20, Infinity, 30);
		});

		it("should fill a rect that extends to outside of the viewbox", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	});

    describe("that fills the entire plane", () => {

		beforeEach(() => {
			contextMock.clear();
			infiniteContext.fillRect(-Infinity, -Infinity, Infinity, Infinity);
		});

		it("should fill the entire viewbox", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});

		describe('and then clears the entire plane', () => {

			beforeEach(() => {
				contextMock.clear();
				infiniteContext.clearRect(-Infinity, -Infinity, Infinity, Infinity)
			})

			it("should have cleared the plane", () => {
				expect(contextMock.getLog()).toMatchSnapshot();
			});
		})
	});

    describe("that fills a rect with negative infinite width and a finite height", () => {

		beforeEach(() => {
			contextMock.clear();
			infiniteContext.fillRect(20, 20, -Infinity, 30);
		});

		it("should fill a rect that extends to outside of the viewbox", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});

		describe.each([
			[10, 10, 40, 40],
			[10, 10, -40, 40],
			[10, 10, Infinity, 40],
			[10, 10, -Infinity, 40],
			[10, 10, 40, Infinity],
			[10, 10, -40, Infinity],
			[10, 10, Infinity, Infinity],
			[10, 10, -Infinity, Infinity],

			[10, 60, 40, -40],
			[10, 60, -40, -40],
			[10, 60, Infinity, -40],
			[10, 60, -Infinity, -40],
			[10, 60, 40, -Infinity],
			[10, 60, -40, -Infinity],
			[10, 60, Infinity, -Infinity],
			[10, 60, -Infinity, -Infinity],
		])("and then clears a rect that intersects the rectangle",(x: number, y: number, w: number, h: number) => {

			beforeEach(() => {
				contextMock.clear();
				infiniteContext.clearRect(x, y, w, h);
			});

			it("should add a clear rect", () => {
				const log: string = contextMock.getLog().join(";");
				expect(log.match(/context\.transform\([^)]+\);context\.clearRect\([^)]+\);context\.restore()/)).toBeTruthy();
			});
		});

		describe.each([
			[30, 10, 40, 40],
			[30, 10, 40, Infinity],
			[30, 10, 40, -Infinity],
			[30, 10, 40, -40],
			[30, 10, Infinity, Infinity],
			[30, 10, Infinity, -Infinity],
			[30, 10, Infinity, 40],
			[30, 10, Infinity, -40],
			[30, 10, -Infinity, -Infinity],
			[30, 10, -Infinity, -10],

			[10, 10, 40, -40],
			[10, 10, 40, -Infinity],
			[10, 10, Infinity, -Infinity],
			[10, 10, Infinity, -40],
			[10, 10, -Infinity, -Infinity],
			[10, 10, -Infinity, -40],

			[10, 60, 40, Infinity],
			[10, 60, Infinity, Infinity],
			[10, 60, -Infinity, Infinity],
			[10, 60, Infinity, 10],
			[10, 60, -Infinity, 10],
		])("and then clears a rect that does not intersect the rectangle",(x: number, y: number, w: number, h: number) => {

			beforeEach(() => {
				contextMock.clear();
				infiniteContext.clearRect(x, y, w, h);
			});

			it("should do nothing", () => {
				expect(contextMock.getLog()).toEqual([]);
			});
		});
	});

    describe("that fills a rect without area", () => {

		beforeEach(() => {
			contextMock.clear();
			infiniteContext.fillRect(-Infinity, 50, 50, 50);
		});

		it("should do nothing", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	});

    describe("that strokes a rect without area", () => {

		beforeEach(() => {
			contextMock.clear();
			infiniteContext.strokeRect(-Infinity, 50, 50, 50);
		});

		it("should do nothing", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	});

    describe("that strokes the entire plane", () => {

		beforeEach(() => {
			contextMock.clear();
			infiniteContext.strokeRect(-Infinity, -Infinity, Infinity, Infinity);
		});

		it("should do nothing", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	});
})