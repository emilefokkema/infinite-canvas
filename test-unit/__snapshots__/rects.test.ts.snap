// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`an infinite canvas context > that begins a path > and then adds a rect that has no area > and then adds a line to a finite point and strokes it > should have drawn a ray from the position of the rect without area 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.beginPath()",
  "context.moveTo(-1.41421356,100)",
  "context.lineTo(100,100)",
  "context.lineWidth = 1",
  "context.stroke()",
]
`;

exports[`an infinite canvas context > that fills a rect with finite width and negative infinite height > should fill a rect that extends to outside of the viewbox 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(20,20)",
  "context.lineTo(50,20)",
  "context.lineTo(50,0)",
  "context.lineTo(20,0)",
  "context.closePath()",
  "context.moveTo(20,20)",
  "context.fill()",
]
`;

exports[`an infinite canvas context > that fills a rect with finite width and positive infinite height > should fill a rect that extends to outside of the viewbox 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(20,20)",
  "context.lineTo(50,20)",
  "context.lineTo(50,200)",
  "context.lineTo(20,200)",
  "context.closePath()",
  "context.moveTo(20,20)",
  "context.fill()",
]
`;

exports[`an infinite canvas context > that fills a rect with negative infinite width and a finite height > should fill a rect that extends to outside of the viewbox 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(20,20)",
  "context.lineTo(0,20)",
  "context.lineTo(0,50)",
  "context.lineTo(20,50)",
  "context.closePath()",
  "context.moveTo(20,20)",
  "context.fill()",
]
`;

exports[`an infinite canvas context > that fills a rect with only a right edge > should fill a rect that fills the left half of the viewbox 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(100,0)",
  "context.lineTo(0,0)",
  "context.lineTo(0,200)",
  "context.lineTo(100,200)",
  "context.lineTo(100,0)",
  "context.lineTo(100,0)",
  "context.closePath()",
  "context.fill()",
]
`;

exports[`an infinite canvas context > that fills a rect with only a top edge > should fill a rect that fills the bottom half of the viewbox 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(0,100)",
  "context.lineTo(0,100)",
  "context.lineTo(200,100)",
  "context.lineTo(200,200)",
  "context.lineTo(0,200)",
  "context.lineTo(0,100)",
  "context.closePath()",
  "context.fill()",
]
`;

exports[`an infinite canvas context > that fills a rect with positive infinite width and a finite height > should fill a rect that extends to outside of the viewbox 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(20,20)",
  "context.lineTo(200,20)",
  "context.lineTo(200,50)",
  "context.lineTo(20,50)",
  "context.closePath()",
  "context.moveTo(20,20)",
  "context.fill()",
]
`;

exports[`an infinite canvas context > that fills a rect without area > should do nothing 1`] = `[]`;

exports[`an infinite canvas context > that fills a rect without bottom edge with negative infinite height and a finite width > should fill a rect that extends to the top and bottom of the viewbox 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(50,200)",
  "context.lineTo(50,0)",
  "context.lineTo(50,0)",
  "context.lineTo(20,0)",
  "context.lineTo(20,0)",
  "context.lineTo(20,200)",
  "context.closePath()",
  "context.fill()",
]
`;

exports[`an infinite canvas context > that fills a rect without left edge with positive infinite width and a finite height > should fill a rect that extends to the left and right of the viewbox 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(0,20)",
  "context.lineTo(0,20)",
  "context.lineTo(200,20)",
  "context.lineTo(200,50)",
  "context.lineTo(0,50)",
  "context.lineTo(0,50)",
  "context.closePath()",
  "context.fill()",
]
`;

exports[`an infinite canvas context > that fills a rect without right edge with negative infinite width and a finite height > should fill a rect that extends to the left and right of the viewbox 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(200,20)",
  "context.lineTo(0,20)",
  "context.lineTo(0,20)",
  "context.lineTo(0,50)",
  "context.lineTo(0,50)",
  "context.lineTo(200,50)",
  "context.closePath()",
  "context.fill()",
]
`;

exports[`an infinite canvas context > that fills a rect without top edge with positive infinite height and a finite width > should fill a rect that extends to the top and bottom of the viewbox 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(50,0)",
  "context.lineTo(50,0)",
  "context.lineTo(50,200)",
  "context.lineTo(20,200)",
  "context.lineTo(20,0)",
  "context.lineTo(20,0)",
  "context.closePath()",
  "context.fill()",
]
`;

exports[`an infinite canvas context > that fills the entire plane > and then clears the entire plane > should have cleared the plane 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
]
`;

exports[`an infinite canvas context > that fills the entire plane > should fill the entire viewbox 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.save()",
  "context.setTransform(1,0,0,1,0,0)",
  "context.rect(0,0,200,200)",
  "context.restore()",
  "context.fill()",
]
`;

exports[`an infinite canvas context > that strokes a rect without area > should do nothing 1`] = `[]`;

exports[`an infinite canvas context > that strokes the entire plane > should do nothing 1`] = `[]`;
